paddle/fluid/operators/cross_entropy2_op.h

```diff
@@ -0,0 +1,188 @@
+
+#pragma once
+
+#include <cmath>
+#include "paddle/fluid/framework/eigen.h"
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/operators/math/cross_entropy.h"
+#include "paddle/fluid/operators/math/math_function.h"
+#include "paddle/fluid/platform/for_range.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+HOSTDEVICE inline platform::float16 RealLog(platform::float16 x) {
+#ifdef __NVCC__
+  return static_cast<platform::float16>(logf(static_cast<float>(x)));
+#else
+  return static_cast<platform::float16>(std::log(static_cast<float>(x)));
+#endif
+}
+
+HOSTDEVICE inline float RealLog(float x) {
+#ifdef __NVCC__
+  return logf(x);
+#else
+  return std::log(x);
+#endif
+}
+
+HOSTDEVICE inline double RealLog(double x) {
+#ifdef __NVCC__
+  return log(x);
+#else
+  return std::log(x);
+#endif
+}
+
+HOSTDEVICE inline platform::float16 RealExp(platform::float16 x) {
+#ifdef __NVCC__
+  return static_cast<platform::float16>(expf(static_cast<float>(x)));
+#else
+  return static_cast<platform::float16>(std::exp(static_cast<float>(x)));
+#endif
+}
+
+HOSTDEVICE inline float RealExp(float x) {
+#ifdef __NVCC__
+  return expf(x);
+#else
+  return std::exp(x);
+#endif
+}
+
+HOSTDEVICE inline double RealExp(double x) {
+#ifdef __NVCC__
+  return exp(x);
+#else
+  return std::exp(x);
+#endif
+}
+
+template <typename T>
+struct CrossEntropyForwardFunctor {
+  CrossEntropyForwardFunctor(const T *x, T *y, const int64_t *label,
+                             int64_t ignore_index, int64_t feature_size)
+      : x_(x),
+        y_(y),
+        label_(label),
+        ignore_index_(ignore_index),
+        feature_size_(feature_size) {}
+
+  HOSTDEVICE void operator()(int64_t row_idx) const {
+    auto col_idx = label_[row_idx];
+    if (col_idx != ignore_index_) {
+      y_[row_idx] = -math::TolerableValue<T>()(
+          RealLog(x_[row_idx * feature_size_ + col_idx]));
+    } else {
+      y_[row_idx] = 0;
+    }
+  }
+
+  const T *x_;
+  T *y_;
+  const int64_t *label_;
+  int64_t ignore_index_;
+  int64_t feature_size_;
+};
+
+template <typename T>
+struct CrossEntropyBackwardFunctor {
+  CrossEntropyBackwardFunctor(T *dx, const T *y, const T *dy,
+                              const int64_t *label, int64_t ignore_index,
+                              int64_t feature_size)
+      : dx_(dx),
+        y_(y),
+        dy_(dy),
+        label_(label),
+        ignore_index_(ignore_index),
+        feature_size_(feature_size) {}
+
+  HOSTDEVICE void operator()(int64_t idx) const {
+    auto row_idx = idx / feature_size_;
+    auto col_idx = idx % feature_size_;
+    auto label = label_[row_idx];
+    if (label == col_idx && label != ignore_index_) {
+      dx_[idx] = -dy_[row_idx] * RealExp(y_[row_idx]);
+    } else {
+      dx_[idx] = 0;
+    }
+  }
+
+  T *dx_;
+  const T *y_;
+  const T *dy_;
+  const int64_t *label_;
+  int64_t ignore_index_;
+  int64_t feature_size_;
+};
```

Comment:
The above code should be placed to `cross_entropy.h`.

Reply:
Done. Move to cross_entropy_op_base.h